Fixes tags on ordered pizza

cacogen opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

 ### Box tags
- Fixes bug where box tags weren't updating correctly on creation
- Adds box tag flavours for the pizzas without them (i.e. Donk Pocket -> Bangin' Donk, Dank -> Fresh Herb, Sassysage -> Sausage Lovers, Arnold -> 9mm Pepperoni)
### Randomised pizza crates
- Pizza crates can now come with any five pizzas, weighted by disruptiveness
- Small chance of getting one pizza bomb in a pizza crate (either armed or not) per shift
### Armed pizza bomb
  - Adds an armed variant of the bomb pizza box, which has the boxtag "Meat Explosion", contains a meat pizza and explodes 5 seconds after opening
  - Makes the pizza party spawner use the armed bomb pizza box, instead of giving whoever a free syndicate pizza bomb they probably aren't allowed to use anyway
 ### Secure kitchen crate
  - Adds the secure kitchen crate from my previous PR
  - Adds a secure pizza crate variant to KiloStation's cargo warehouse where the freezer crate was, which contains the pizza party and is locked with kitchen access
 ### Ghost examines
- Adds ghost examine messages for pizza boxes with bombs or the nanomachine pizza in them

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

Fixes a bug. Improves the code. Makes pizza boxes better.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl: cacogen
tweak: Cargo pizza crate now picks randomly from a list of most pizzas (non-repeating)
fix: Cargo pizza crate pizza boxes now display their tags correctly
add: New box tags for pizzas that previously lacked them
add: Pizza bomb variant that starts armed (look for the "Meat Explosion" box tag)
balance: Cargo pizza crates have a 15% chance to spawn a pizza bomb (armed or a free one) once a round
balance: The pizza bomb the pizza party spawner may create uses the armed variant now
add: Adds a secure pizza crate to the Ice Box cargo warehouse (kitchen access, may contain an armed pizza bomb)
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

cacogen committed:
Fixes tags on ordered pizza

Adds secure kitchen crate

cacogen committed:
Fixes tags on ordered pizza

Adds secure kitchen crate

Tlaltecuhtli commented:
cm has id locked pill bottle, tg has id locked pizza boxes

cacogen commented:
Yeah it's going in

cacogen committed:
Additions and improvements

- Adds an armed bomb pizza box
- Improves code in pizzabox.dm
- Adds ghost examine messages for pizza boxes with bombs or the nanomachine pizza
- Makes the pizza party spawner use the armed pizza box bomb (not what crates use)
- Adds box tag flavours for the pizzas without them

cacogen committed:
Additions and improvements

- Adds an armed bomb pizza box
- Improves code in pizzabox.dm
- Adds ghost examine messages for pizza boxes with bombs or the nanomachine pizza
- Makes the pizza party spawner use the armed pizza box bomb (not what crates use)
- Adds box tag flavours for the pizzas without them

cacogen committed:
We call them chilly bins

Due to sun exposure

cacogen committed:
We call them chilly bins

Due to sun exposure

cacogen committed:
Merge remote-tracking branch 'origin/master' into fixpizzatags

cacogen committed:
Meat Explosion

cacogen committed:
Meat Explosion

cacogen committed:
Makes it so pizza crates can spawn any pizza

- Updates pizza crate desc to reflect this randomness
- Gives the infinite pizza box the "Your Favourite" tag by default (the damage flags give it away, and this makes you understand why it's giving you the same pizza every time)
- Tweaks observer descriptions for bomb and nanite pizza boxes
- Clarifies var used by infinite pizza boxes

cacogen committed:
Makes it so pizza crates can spawn any pizza

- Updates pizza crate desc to reflect this randomness
- Gives the infinite pizza box the "Your Favourite" tag by default (the damage flags give it away, and this makes you understand why it's giving you the same pizza every time)
- Tweaks observer descriptions for bomb and nanite pizza boxes
- Clarifies var used by infinite pizza boxes

cacogen committed:
Pizza crates have a small chance per pizza to spawn a bomb pizza

- Makes probability numbers into vars
- Autodocs vars
- Makes dank pizza more likely because it's not very disruptive to the rounds of people who eat it
- Makes it so anomalous pizza boxes always have the correct box tag for the person examining them
- Tweaks minor shit

cacogen committed:
Pizza crates have a small chance per pizza to spawn a bomb pizza

- Makes probability numbers into vars
- Autodocs vars
- Makes dank pizza more likely because it's not very disruptive to the rounds of people who eat it
- Makes it so anomalous pizza boxes always have the correct box tag for the person examining them
- Tweaks minor shit

Rohesie commented on _maps/map_files/IceBoxStation/IcemoonUnderground_Above.dmm:
```diff
@@ -331,9 +331,15 @@
-/obj/structure/closet/crate/freezer,
 /obj/machinery/power/apc/auto_name/east,
 /obj/structure/cable,
+/obj/structure/closet/crate/secure/freezer{
+	desc = "An insulated crate with a lock on it, used to secure pizza.";
+	name = "secure pizza crate";
+	req_access = list(28);
+	tamperproof = 1
+	},
```
Can you make an object variant instead of var-editing this?
Especially the tamperproof variable.

Rohesie commented on code/game/objects/items/food/pizza.dm:
```diff
@@ -8,6 +8,8 @@
 	/// type is spawned 6 at a time and replaces this pizza when processed by cutting tool
 	var/obj/item/food/pizzaslice/slice_type
+	///What label pizza boxes use if this pizza spawns in them.
+	var/boxtag
```
```suggestion
	var/boxtag = ""
```

Rohesie commented on code/modules/cargo/packs.dm:
```diff
@@ -1807,31 +1807,84 @@
+
+		//here we randomly replace our pizzas for a chance at the full range
+		var/obj/item/food/pizza/replacement_type = pickweight(pizza_types)
+		pizza_types -= replacement_type
```
Why are you removing the type? Will you be using this list anywhere else?

cacogen replied:
To prevent duplicates. We either already have that pizza, or we're about to create it.

Rohesie replied:
Aha, didn't notice it was inside a loop

Rohesie commented on code/modules/cargo/packs.dm:
```diff
@@ -1807,31 +1807,84 @@
+
+	var/list/pizza_types = list(
+		/obj/item/food/pizza/margherita = 1,
+		/obj/item/food/pizza/margherita/robo = 0.1,
```
Not really a fan of this.

cacogen replied:
I'll get rid of it

Rohesie commented on code/modules/food_and_drinks/pizzabox.dm:
```diff
@@ -271,42 +285,39 @@
-/obj/item/pizzabox/margherita/proc/AddPizza()
-	pizza = new /obj/item/food/pizza/margherita(src)
+/obj/item/pizzabox/margherita
+	pizza = new /obj/item/food/pizza/margherita
```
Keep the type path as a variable, and create it during initialization instead.

Rohesie commented on code/modules/food_and_drinks/pizzabox.dm:
```diff
@@ -323,35 +334,46 @@
 		/obj/item/food/pizza/dank = 0.1,
 	) //pizzas here are weighted by chance to be someone's favorite
 	var/static/list/pizza_preferences
+	boxtag = "Your Favourite"
+	var/mob/living/carbon/human/attuned_to
```
* Put redefined variables (the ones without `var/`) over new ones (the ones with) for organization purposes.
* Document new variables.
* Clear dangling references (register to their destruction and handle that case).

cacogen replied:
I did the first two. I'm not sure what the third bullet point means.

Rohesie commented on code/modules/food_and_drinks/pizzabox.dm:
```diff
@@ -323,35 +334,46 @@
 	if(!pizza_preferences)
 		pizza_preferences = list()
 
 /obj/item/pizzabox/infinite/examine(mob/user)
+	if(!open && ishuman(user))
+		attune_pizza(user) //this is so our boxtag is always correct
```
Can you explain what you are doing here?

cacogen replied:
I explained it with a better comment

cacogen committed:
Instantiates boxtag as an empty string

Co-authored-by: Rohesie <rohesie@gmail.com>

cacogen committed:
Merge remote-tracking branch 'cacogen/fixpizzatags' into fixpizzatags

cacogen committed:
Removes nanite pizza from pizza crate chances

cacogen committed:
Changes for Rohesie

cacogen commented:
I don't know why it has the sound, config and UI labels.

Rohesie commented on code/modules/food_and_drinks/pizzabox.dm:
```diff
@@ -44,6 +48,9 @@
 /obj/item/pizzabox/update_icon()
 	// Description
 	desc = initial(desc)
+	if(pizza && pizza.boxtag && !boxtag_set)
```
```suggestion
	if(pizza?.boxtag && !boxtag_set)
```

Rohesie commented on code/modules/food_and_drinks/pizzabox.dm:
```diff
@@ -321,37 +334,50 @@
+			pizza = new favourite_pizza_type
+			boxtag_set = FALSE
+			update_icon() //update our boxtag to match our new pizza
+			pizza.foodtypes = nommer.dna.species.liked_food //it's our favorite!
+			attuned_to = nommer
```
What happens if the `nommer` gets qdeleted before the pizza?
Will this variable point to a qdeleted entity, forcing a hard deletion?
Does this variable get cleaned anywhere, such as on the object's `Destroy()`?

If you don't understand how reference handling and hard deletions work, poke me on the coder channel and I'll explain.

cacogen replied:
So I wasn't sure how to clear attuned_to in the event the eater is destroyed (as an aside it seems misguided to me to actually delete a player mob) so I just made it "respawn" the pizza in the box if the type is not the human's favourite or the foodtype isn't what the human likes.

When I tested this with an ash walker and a human, the human didn't get any adverse taste messages but the ash walker got them on both its own pizza and the human's pizza. I don't think I introduced that, though. If not, I may not try to fix it.

cacogen committed:
Removes attuned_to

Rohesie commented on code/modules/food_and_drinks/pizzabox.dm:
```diff
@@ -330,28 +344,37 @@
-	pizza = new pizza_type (src)
-	pizza.foodtypes = noms.dna.species.liked_food //it's our favorite!
+/obj/item/pizzabox/infinite/proc/attune_pizza(mob/living/carbon/human/nommer) //tonight on "proc names I never thought I'd type"
+	if(nommer.ckey)
+		//list our ckey and assign a favourite pizza
```
Do an early return here.

cacogen committed:
Merge remote-tracking branch 'origin/master' into fixpizzatags

cacogen committed:
Early return

- Redoes a whole bunch of logic shit I'm bad at this is this better I don't know why'd I do it

cacogen committed:
ckey's pizza

cacogen committed:
Merge branch 'master' into fixpizzatags

cacogen committed:
Fixes merge conflict

Rohesie commented on code/modules/food_and_drinks/pizzabox.dm:
```diff
@@ -330,28 +344,43 @@
+
+	//list our ckey and assign it a favourite pizza
+	if(!pizza_preferences[nommer.ckey])
+		pizza_preferences[nommer.ckey] = pickweight(pizza_types)
+		if(nommer.has_quirk(/datum/quirk/pineapple_liker))
```
```suggestion
		if(nommer.has_quirk(/datum/quirk/pineapple_liker))
```

Rohesie commented on code/modules/food_and_drinks/pizzabox.dm:
```diff
@@ -330,28 +344,43 @@
-			pizza_preferences[noms.ckey] = /obj/item/food/pizza/dank
+			pizza_preferences[nommer.ckey] = pickweight(pineapple_pizza_liker)
+		else if(nommer.mind && nommer.mind.assigned_role == "Botanist")
+			pizza_preferences[nommer.ckey] = /obj/item/food/pizza/dank
+
```
```suggestion
			pizza_preferences[nommer.ckey] = /obj/item/food/pizza/dank
		else
			pizza_preferences[nommer.ckey] = pickweight(pizza_types)
```

Rohesie reviewed (changes requested):
Confirm when this is tested locally and the change requests made. Should be ready for merge.

cacogen committed:
Apply suggestions from code review

Co-authored-by: Rohesie <rohesie@gmail.com>

cacogen committed:
Fixes inadvertent runtime probably caused by pickweight()

cacogen commented:
@Rohesie Tested your changes then, made an adjustment to avoid a runtime. I've tested everything else extensively so in theory it should work without any problems.

Rohesie commented on code/modules/food_and_drinks/pizzabox.dm:
```diff
@@ -330,28 +344,43 @@
+		else
+			pizza.foodtypes = nommer.dna.species.liked_food //make sure it's our favourite
+			return
+	if(pizza_preferences[nommer.ckey]) //make sure it hasn't been set to null, which happens sometimes
```
This could result on a `null` `pizza` if `pickweight()` returns `null`, no?
The check should be put a bit before with an early return.

cacogen replied:
It can result in a null pizza. I can early return but the result is an empty pizza box, which isn't optimal but isn't harmful.

Rohesie replied:
Correct me if I'm wrong, but doesn't it go this way?

- `if(!pizza_preferences[nommer.ckey])` is evaluated to false
- all the ifs inside of it evaluate to false, falling into the last one
- `pizza_preferences[nommer.ckey] = pickweight(pizza_types)` happens, but `pizza_type` has a length of zero by now
- `if(pizza)` is evaluated to true, and thus `if(pizza.type != pizza_preferences[nommer.ckey])` passes as the type is not -`null`
- `QDEL_NULL(pizza)` happens, and the box now has a `null` pizza
- `if(pizza_preferences[nommer.ckey])` evaluates to false, as this is `null
- now the box has no pizza

cacogen replied:
Yeah, it can result in no pizza. So I added that check to prevent it trying to make a null pizza, which created a runtime.

Gameplay-wise, you open the box, it's empty and the incorrect boxtag can't be seen. Chances are extremely high closing and reopening it once will pick a pizza. But eventually closing and reopening it will fix it.

I can't think of a more optimal outcome (aside from fixing pick_weight() which I doubt I could figure out) and I don't understand why the early return would be better. Which is probably reflective of my coding knowledge.

The other thing I could do is loop up to three times if pick_weight() hasn't done its job to make it almost certain pick_weight() will work but I don't know if that would be acceptable in terms of performance and time taken to open it.

Rohesie replied:
First things first: why is `pizza_types` empty sometimes? Do you know that?
Or are you saying that `pickweight()` is returning `null` on a non-empty list?

cacogen replied:
I don't know for certain but I'm pretty sure it's pickweight() causing the null. There's a comment on the second pickweight() proc (the one that allows null) mentioning that it happens sometimes.

Rohesie replied:
Add debugging lines when testing.
Things like
```dm
to_chat(world, "pizza_types length: [length(pizza_types)]")
```
Add `stack_traces()` if necessary.

Confirm `pickweight()` is outputting a `null` on a list with elements, and once you do, check what kind of contents the list has:
```dm
for(i in length(pizza_types))
	var/pizza_key = pizza_types[i]
	to_chat(world, "pizza_types [i]: [pizza_key] = [pizza_types[pizza_key]]")
```

cacogen replied:
The code
```dm
for(var/i in 1 to 100)
	var/pizza_type_to_spawn = pickweight(pizza_types)
	if(pizza_type_to_spawn)
		to_chat(world, "pickweight() picked [pizza_type_to_spawn]!")
		//var/obj/item/food/pizza/spawned_pizza = new pizza_type_to_spawn(user.loc)
	else
		to_chat(world, "pickweight() returned null!")
		for(var/e in 1 to length(pizza_types))
			var/pizza_key = pizza_types[e]
			to_chat(world, "pizza_types [e]: [pizza_key] = [pizza_types[pizza_key]]")
```

The result

```dm
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() returned null!
pizza_types 1: /obj/item/food/pizza/meat = 1
pizza_types 2: /obj/item/food/pizza/mushroom = 1
pizza_types 3: /obj/item/food/pizza/margherita = 1
pizza_types 4: /obj/item/food/pizza/sassysage = 0.8
pizza_types 5: /obj/item/food/pizza/vegetable = 0.8
pizza_types 6: /obj/item/food/pizza/pineapple = 0.5
pizza_types 7: /obj/item/food/pizza/donkpocket = 0.3
pizza_types 8: /obj/item/food/pizza/dank = 0.1
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() returned null!
pizza_types 1: /obj/item/food/pizza/meat = 1
pizza_types 2: /obj/item/food/pizza/mushroom = 1
pizza_types 3: /obj/item/food/pizza/margherita = 1
pizza_types 4: /obj/item/food/pizza/sassysage = 0.8
pizza_types 5: /obj/item/food/pizza/vegetable = 0.8
pizza_types 6: /obj/item/food/pizza/pineapple = 0.5
pizza_types 7: /obj/item/food/pizza/donkpocket = 0.3
pizza_types 8: /obj/item/food/pizza/dank = 0.1
pickweight() returned null!
pizza_types 1: /obj/item/food/pizza/meat = 1
pizza_types 2: /obj/item/food/pizza/mushroom = 1
pizza_types 3: /obj/item/food/pizza/margherita = 1
pizza_types 4: /obj/item/food/pizza/sassysage = 0.8
pizza_types 5: /obj/item/food/pizza/vegetable = 0.8
pizza_types 6: /obj/item/food/pizza/pineapple = 0.5
pizza_types 7: /obj/item/food/pizza/donkpocket = 0.3
pizza_types 8: /obj/item/food/pizza/dank = 0.1
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() returned null!
pizza_types 1: /obj/item/food/pizza/meat = 1
pizza_types 2: /obj/item/food/pizza/mushroom = 1
pizza_types 3: /obj/item/food/pizza/margherita = 1
pizza_types 4: /obj/item/food/pizza/sassysage = 0.8
pizza_types 5: /obj/item/food/pizza/vegetable = 0.8
pizza_types 6: /obj/item/food/pizza/pineapple = 0.5
pizza_types 7: /obj/item/food/pizza/donkpocket = 0.3
pizza_types 8: /obj/item/food/pizza/dank = 0.1
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() returned null!
pizza_types 1: /obj/item/food/pizza/meat = 1
pizza_types 2: /obj/item/food/pizza/mushroom = 1
pizza_types 3: /obj/item/food/pizza/margherita = 1
pizza_types 4: /obj/item/food/pizza/sassysage = 0.8
pizza_types 5: /obj/item/food/pizza/vegetable = 0.8
pizza_types 6: /obj/item/food/pizza/pineapple = 0.5
pizza_types 7: /obj/item/food/pizza/donkpocket = 0.3
pizza_types 8: /obj/item/food/pizza/dank = 0.1
pickweight() returned null!
pizza_types 1: /obj/item/food/pizza/meat = 1
pizza_types 2: /obj/item/food/pizza/mushroom = 1
pizza_types 3: /obj/item/food/pizza/margherita = 1
pizza_types 4: /obj/item/food/pizza/sassysage = 0.8
pizza_types 5: /obj/item/food/pizza/vegetable = 0.8
pizza_types 6: /obj/item/food/pizza/pineapple = 0.5
pizza_types 7: /obj/item/food/pizza/donkpocket = 0.3
pizza_types 8: /obj/item/food/pizza/dank = 0.1
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() returned null!
pizza_types 1: /obj/item/food/pizza/meat = 1
pizza_types 2: /obj/item/food/pizza/mushroom = 1
pizza_types 3: /obj/item/food/pizza/margherita = 1
pizza_types 4: /obj/item/food/pizza/sassysage = 0.8
pizza_types 5: /obj/item/food/pizza/vegetable = 0.8
pizza_types 6: /obj/item/food/pizza/pineapple = 0.5
pizza_types 7: /obj/item/food/pizza/donkpocket = 0.3
pizza_types 8: /obj/item/food/pizza/dank = 0.1
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() returned null!
pizza_types 1: /obj/item/food/pizza/meat = 1
pizza_types 2: /obj/item/food/pizza/mushroom = 1
pizza_types 3: /obj/item/food/pizza/margherita = 1
pizza_types 4: /obj/item/food/pizza/sassysage = 0.8
pizza_types 5: /obj/item/food/pizza/vegetable = 0.8
pizza_types 6: /obj/item/food/pizza/pineapple = 0.5
pizza_types 7: /obj/item/food/pizza/donkpocket = 0.3
pizza_types 8: /obj/item/food/pizza/dank = 0.1
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() returned null!
pizza_types 1: /obj/item/food/pizza/meat = 1
pizza_types 2: /obj/item/food/pizza/mushroom = 1
pizza_types 3: /obj/item/food/pizza/margherita = 1
pizza_types 4: /obj/item/food/pizza/sassysage = 0.8
pizza_types 5: /obj/item/food/pizza/vegetable = 0.8
pizza_types 6: /obj/item/food/pizza/pineapple = 0.5
pizza_types 7: /obj/item/food/pizza/donkpocket = 0.3
pizza_types 8: /obj/item/food/pizza/dank = 0.1
```

cacogen replied:
What I was suggesting to work around pick_weight() was something akin to the following:

```dm
/obj/item/pizzabox/floorpizza_spawner/attack_self(mob/living/user)
	for(var/i in 1 to 1000)
		var/pizza_type_to_spawn = pickweight(pizza_types)
		if(!pizza_type_to_spawn)
			for(var/e in 1 to 3)
				pizza_type_to_spawn = pickweight(pizza_types)
				if(pizza_type_to_spawn)
					break
		if(pizza_type_to_spawn)
			var/obj/item/food/pizza/spawned_pizza = new pizza_type_to_spawn(user.loc)
		else
			to_chat(world,"Loop [i] of 1000: No pizza type to spawn!")
```

I spawned 20,000 pizzas with this method and only two iterations ended with a null pizza type.

> Loop 852 of 1000: No pizza type to spawn!
> Loop 202 of 1000: No pizza type to spawn!

Rohesie replied:
I see the issue. `pickweight()` uses `rand()`, which uses integers, not floats.
Try multiplying each element of your list by 10 so you have no decimal values, and try again.
(ie: `/obj/item/food/pizza/pineapple = 5` instead of `= 0.5` )

cacogen replied:
That seemed to fix it. Spawned another 20k pizzas without a single null.

cacogen committed:
Multiplies by 10 to prevent getting a null

cacogen committed:
Merge remote-tracking branch 'origin/master' into fixpizzatags

cacogen committed:
Removes check for pizza_preferences[nommer.ckey]

cacogen commented:
@Rohesie Is this okay now?

Rohesie commented:
If it's tested locally and working properly then set it again for review. Should be ready now.

Rohesie commented on code/modules/cargo/packs.dm:
```diff
@@ -1785,31 +1785,81 @@
-			if(prob(1)) //1% chance for each box, so 4% total chance per order
+
+	var/list/pizza_types = list(
+		/obj/item/food/pizza/margherita = 1,
+		/obj/item/food/pizza/meat = 1,
+		/obj/item/food/pizza/mushroom = 1,
+		/obj/item/food/pizza/vegetable = 1,
+		/obj/item/food/pizza/donkpocket = 1,
+		/obj/item/food/pizza/dank = 0.7,
+		/obj/item/food/pizza/sassysage = 1,
+		/obj/item/food/pizza/pineapple = 1,
+		/obj/item/food/pizza/arnold = 0.3
```
Same as the other, multiply everyone of this by 10. We want integers here.

cacogen replied:
Damn, forgot about this. Thanks.

Rohesie commented:
Actually do that change and then test it, first.

cacogen committed:
Fixes weighted list on pizza crate

- Adds note to pickweight() proc saying not to use decimals with it

cacogen commented:
Okay, I made the changes and I used an express cargo console to spawn a few pizza crates to test them. All the boxes were filled and the tags were correct. I'm guessing it should be fine. But if there are any bugs I encounter playtesting this on the server or that get reported I'll fix them.
